framework/src/main/java/org/tron/core/metrics/blockchain/BlockChainMetricManager.java

```diff
@@ -0,0 +1,217 @@
+    }
+    witnessInfo.put(witnessAddress, block);
+
+    //latency
+    long netTime = nowTime - block.getTimeStamp();
+    metricsService.histogramUpdate(MetricsKey.NET_BLOCK_LATENCY, netTime);
+    metricsService.histogramUpdate(MetricsKey.NET_BLOCK_LATENCY_WITNESS + witnessAddress, netTime);
+    if (netTime >= 1000) {
+      metricsService.counterInc(MetricsKey.NET_BLOCK_LATENCY + ".1S", 1L);
+      metricsService.counterInc(MetricsKey.NET_BLOCK_LATENCY_WITNESS + witnessAddress + ".1S", 1L);
+      if (netTime >= 2000) {
+        metricsService.counterInc(MetricsKey.NET_BLOCK_LATENCY + ".2S", 1L);
+        metricsService.counterInc(MetricsKey.NET_BLOCK_LATENCY_WITNESS + witnessAddress + ".2S",
+                1L);
+        if (netTime >= 3000) {
+          metricsService.counterInc(MetricsKey.NET_BLOCK_LATENCY + ".3S", 1L);
+          metricsService.counterInc(MetricsKey.NET_BLOCK_LATENCY_WITNESS + witnessAddress + ".3S",
+                  1L);
+        }
+      }
+    }
+
+    //TPS
+    if (block.getTransactions().size() > 0) {
+      metricsService.meterMark(MetricsKey.BLOCKCHAIN_TPS, block.getTransactions().size());
+    }
+  }
+
+  private List<WitnessInfo> getSrList() {
+    List<WitnessInfo> witnessInfos = new ArrayList<>();
+
+    List<ByteString> witnessList = chainBaseManager.getWitnessScheduleStore().getActiveWitnesses();
+    for (ByteString witnessAddress : witnessList) {
+      String address = Hex.toHexString(witnessAddress.toByteArray());
+      if (witnessInfo.containsKey(address)) {
+        BlockCapsule block = witnessInfo.get(address);
+        WitnessInfo witness = new WitnessInfo(address,
+                block.getInstance().getBlockHeader().getRawData().getVersion());
+        witnessInfos.add(witness);
+      }
+    }
+    return witnessInfos;
+  }
+
+  private RateInfo getBlockProcessTime() {
+    RateInfo blockProcessTime = new RateInfo();
+    blockProcessTime.setCount(metricsService.getMeter(MetricsKey.BLOCKCHAIN_BLOCK_COUNT)
+            .getCount());
+    blockProcessTime.setMeanRate(getAvgBlockProcessTimeByGap(0));
+    blockProcessTime.setOneMinuteRate(getAvgBlockProcessTimeByGap(1));
+    blockProcessTime.setFiveMinuteRate(getAvgBlockProcessTimeByGap(5));
+    blockProcessTime.setFifteenMinuteRate(getAvgBlockProcessTimeByGap(15));
+    return blockProcessTime;
+  }
+
+  private RateInfo getTransactionRate() {
+    Meter transactionRate = metricsService.getMeter(MetricsKey.BLOCKCHAIN_TPS);
+    RateInfo tpsInfo = new RateInfo();
+    tpsInfo.setCount(transactionRate.getCount());
+    tpsInfo.setMeanRate(transactionRate.getMeanRate());
+    tpsInfo.setOneMinuteRate(transactionRate.getOneMinuteRate());
+    tpsInfo.setFiveMinuteRate(transactionRate.getFiveMinuteRate());
+    tpsInfo.setFifteenMinuteRate(transactionRate.getFifteenMinuteRate());
+    return tpsInfo;
+  }
+
+
+  // gap: 1 minute, 5 minute, 15 minute, 0: avg for total block and time
+  private double getAvgBlockProcessTimeByGap(int gap) {
```

Comment:
no need conversion, This function is useless. delete it

Reply:
no. This function is needed in order to keep more precious avg block process time